llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -769,14 +783,27 @@ NVPTXTargetLowering::NVPTXTargetLowering(const NVPTXTargetMachine &TM,
   // promoted to f32. v2f16 is expanded to f16, which is then promoted
   // to f32.
   for (const auto &Op :
-       {ISD::FDIV, ISD::FREM, ISD::FSQRT, ISD::FSIN, ISD::FCOS, ISD::FABS}) {
+       {ISD::FDIV, ISD::FREM, ISD::FSQRT, ISD::FSIN, ISD::FCOS}) {
     setOperationAction(Op, MVT::f16, Promote);
     setOperationAction(Op, MVT::bf16, Promote);
     setOperationAction(Op, MVT::f32, Legal);
     setOperationAction(Op, MVT::f64, Legal);
     setOperationAction(Op, MVT::v2f16, Expand);
     setOperationAction(Op, MVT::v2bf16, Expand);
+    if (getOperationAction(Op, MVT::bf16) == Promote)
+      AddPromotedToType(Op, MVT::bf16, MVT::f32);
+  }
+  for (const auto &Op : {ISD::FABS}) {
+    setOperationAction(Op, MVT::f16, Promote);
+    setBF16OperationAction(Op, MVT::bf16, Legal, Promote);
+    setOperationAction(Op, MVT::f32, Legal);
+    setOperationAction(Op, MVT::f64, Legal);
+    setOperationAction(Op, MVT::v2f16, Expand);
+    setBF16OperationAction(Op, MVT::v2bf16, Legal, Expand);
+    if (getOperationAction(Op, MVT::bf16) == Promote)
```

Comment:
Reasonable. But for now if AddPromotedToType is not declared, it would raise an error: 

Trying to promote node
llc: /xxx/llvm-project/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:5658: llvm::SDValue llvm::SelectionDAG::getNode(unsigned int, const llvm::SDLoc&, llvm::EVT, llvm::SDValue, llvm::SDNodeFlags): Assertion `N1.getValueType().bitsLT(VT) && "Invalid fpext node, dst < src!"' failed.

Reply:
I read the code and figured out the reason why autopromote not working:

In function [getTypeToPromoteTo()](https://github.com/llvm/llvm-project/blob/f2517cbceec0bd9b049ba24f36cb3a2508c988fa/llvm/include/llvm/CodeGen/TargetLowering.h#L1567) of `TargetLowering.h`, the autopromote strategy iteratively try
`NVT = (MVT::SimpleValueType)(NVT.SimpleTy+1);`until `isTypeLegal(NVT)` returns true. 

In `llvm/CodeGen/GenVT.inc` the enum declaration of floating type is `bf16 -> f16->f32->f64->f80`, thus `bf16` would fallback to f16 in this rule, which is not what we expects.

Reply:
Another implementation possibility is to add a special case in [getTypeToPromoteTo()](https://github.com/llvm/llvm-project/blob/f2517cbceec0bd9b049ba24f36cb3a2508c988fa/llvm/include/llvm/CodeGen/TargetLowering.h#L1567) to promote bf16 to f32, the advantage is to declare it for all Ops, while might not be that gentle as well.

Reply:
> the enum declaration of floating type is bf16 -> f16->f32->f64->f80, thus bf16 would fallback to f16 in this rule, which is not what we expects.

Yeah. Both bf16 and f16 should be promoted to f32. We should fix that in a separate patch.

Explicitly specifying promotion type is OK, but this looks like a generic issue that should be fixed for everyone.

Reply:
Agree